### PR TITLE
bpo-41100: ctypes fixes for arm64 Mac OS

### PR DESCRIPTION
**Pull request by lawrence-danna-apple**
Posted in response to this comment by Ronald https://github.com/python/cpython/pull/21242#discussion_r447926261

This PR would supersede these two, combining because the third commit depends on both

* https://github.com/python/cpython/pull/21242
* https://github.com/python/cpython/pull/21239

This has three commits:

* ctypes: use the correct ABI for variadic functions
* use system libffi for Mac OS 10.15 and up
* setup.py: probe  libffi for ffi_closure_alloc and ffi_prep_cif_var

The first two are from those previous PRs.   The third makes `setup.py` probe for those functions by searching libffi's headers.

https://bugs.python.org/issue41100

### CODE REVIEW

**Comment by maxbelanger** [comment]
@lawrence-danna-apple We currently redistribute `libffi` along with Python to support Mac OS X 10.10 as our deployment target. Your change implies that there were changes made to `libffi` (and the copy that ships with macOS 11) for arm64 macOS. Are these upstreamed and available?

**Comment by lawrence-danna-apple** [comment]
> @lawrence-danna-apple We currently redistribute `libffi` along with Python to support Mac OS X 10.10 as our deployment target. Your change implies that there were changes made to `libffi` (and the copy that ships with macOS 11) for arm64 macOS. Are these upstreamed and available?

We also distribute `libffi` as part of Mac OS.   Our changes for 10.15, which cover the necessary API changes are available here:  https://opensource.apple.com/source/libffi/libffi-25/  

I'm not sure what the status is as to upstreaming, but I'll do a review of it and see if anything necessary is still missing from upstream.

Noe that in the changes I'm proposing here, the Mac OS system `libffi` is only used if the deployment target is set to 10.15 or better.

**Comment by maxbelanger** [comment]
> > @lawrence-danna-apple We currently redistribute `libffi` along with Python to support Mac OS X 10.10 as our deployment target. Your change implies that there were changes made to `libffi` (and the copy that ships with macOS 11) for arm64 macOS. Are these upstreamed and available?
> 
> We also distribute `libffi` as part of Mac OS. Our changes for 10.15, which cover the necessary API changes are available here: https://opensource.apple.com/source/libffi/libffi-25/
> 
> I'm not sure what the status is as to upstreaming, but I'll do a review of it and see if anything necessary is still missing from upstream.
> 
> Noe that in the changes I'm proposing here, the Mac OS system `libffi` is only used if the deployment target is set to 10.15 or better.

Got it, thanks for clarifying. It looks like there may be some issues in `libffi` w/ XC12 + arm64: https://github.com/libffi/libffi/issues/571. Good to know that we could rely on `libffi-25`, though.

File: Doc/library/ctypes.rst [hunk 1]
```diff
@@ -1579,6 +1579,13 @@ They are instances of a private class:
       value usable as argument (integer, string, ctypes instance).  This allows
       defining adapters that can adapt custom objects as function parameters.
 
+   .. attribute:: variadic
+
+      Assign a boolean to specify that the function takes a variable nubmer of
```
Comment (archoversight) [inline Doc/library/ctypes.rst]:
```suggestion
      Assign a boolean to specify that the function takes a variable number of
```

**Review by ronaldoussoren — COMMENTED** [review]
> @lawrence-danna-apple We currently redistribute `libffi` along with Python to support Mac OS X 10.10 as our deployment target. Your change implies that there were changes made to `libffi` (and the copy that ships with macOS 11) for arm64 macOS. Are these upstreamed and available?

Isn't it possible/save to use the system libffi on 10.9 or later?  The 10.9 SDK includes libffi, even if it uses an older API (just like libffi_osx in the CPython tree).

I'd prefer to drop the copy of libffi included with CPython and always use the system libffi, without including a copy of libffi in the installers for macOS.

**Comment by ronaldoussoren** [comment]
I'm a bit conflicted about the "variadic" attribute for _FuncPtr.   I agree that it is necessary, but this is an API change and as such is a lot harder to back port.

**Comment by lawrence-danna-apple** [comment]
> I'd prefer to drop the copy of libffi included with CPython and always use the system libffi

I'm not sure if this is possible.   The libffi prior to 10.15 won't have the APIs we need to support arm64.    However, we may be able to use ifdefs and availability to sort it out.  I'll try

**Comment by lawrence-danna-apple** [comment]
@ronaldoussoren 

> I'm a bit conflicted about the "variadic" attribute for _FuncPtr. I agree that it is necessary

Well, it looks like I misread the documentation.   It turns out that in the case where a variadic function is called with zero variadic arguments, the two calling conventions coincide.   So the flag is not necessary after all.

**Comment by lawrence-danna-apple** [comment]
@ronaldoussoren 

I found it is possible to use the system libffi even if the deployment target is set to before 10.15.    I've added runtime checks for `ffi_prep_cif_var`,  `ffi_prep_closure_loc` and `ffi_closure_alloc`.    If they are present, it will use them.   If not, it will fall back on the old way of doing things.

I've also enabled the system ffi if either the deployment target is 10.15 OR the arm64 is in the target arches.     I'm not sure if Mac OS on PPC is still supported by python, but these changes shouldn't break it if it is.   

The only thing that will not work is to build a super-fat binary that works on ppc, x86 and arm64 at the same time.

File: Modules/_ctypes/callbacks.c [hunk 2]
```diff
@@ -408,13 +409,29 @@ CThunkObject *_ctypes_alloc_callback(PyObject *callable,
                      "ffi_prep_cif failed with %d", result);
         goto error;
     }
-#if defined(X86_DARWIN) || defined(POWERPC_DARWIN)
-    result = ffi_prep_closure(p->pcl_write, &p->cif, closure_fcn, p);
+#if HAVE_FFI_PREP_CLOSURE_LOC
+#   if USING_APPLE_OS_LIBFFI
+#      define HAVE_FFI_PREP_CLOSURE_LOC_RUNTIME __builtin_available(macos 11, ios 13, watchos 6, tvos 13, *)
```
Comment (ned-deily) [inline Modules/_ctypes/callbacks.c]: `__builtin_available` is itself not available with older versions of macOS and Xcode, in particular with macOS 10.9 and Xcode 6, which we currently need to use to build our most common installer variants.  And there are other projects like MacPorts that builds Python for older macOS systems.  The general review comment has a link for a possible workaround from the curl project.

File: Modules/_ctypes/callproc.c [hunk 3]
```diff
@@ -835,14 +838,63 @@ static int _call_function_pointer(int flags,
...
-                               cc,
-                               argcount,
-                               restype,
-                               atypes)) {
-        PyErr_SetString(PyExc_RuntimeError,
-                        "ffi_prep_cif failed");
-        return -1;
+
+#   if USING_APPLE_OS_LIBFFI
+#      define HAVE_FFI_PREP_CIF_VAR_RUNTIME __builtin_available(macos 11, ios 13, watchos 6, tvos 13, *)
```
Comment (ned-deily) [inline Modules/_ctypes/callproc.c]: `__builtin_available` again

File: Modules/_ctypes/malloc_closure.c [hunk 4]
```diff
@@ -89,16 +89,27 @@ static void more_core(void)
 /******************************************************************/
 
 /* put the item back into the free list */
-void ffi_closure_free(void *p)
+void Py_ffi_closure_free(void *p)
 {
+#if USING_APPLE_OS_LIBFFI
+    if (__builtin_available(macos 11, ios 13, watchos 6, tvos 13, *)) {
```
Comment (ned-deily) [inline Modules/_ctypes/malloc_closure.c]: `__builtin_available` again

File: Modules/_ctypes/malloc_closure.c [hunk 5]
```diff
@@ -89,16 +89,27 @@ static void more_core(void)
...
     item->next = free_list;
     free_list = item;
 }
 
 /* return one item from the free list, allocating more if needed */
-void *ffi_closure_alloc(size_t ignored, void** codeloc)
+void *Py_ffi_closure_alloc(size_t size, void** codeloc)
 {
+#if USING_APPLE_OS_LIBFFI
+    if (__builtin_available(macos 11, ios 13, watchos 6, tvos 13, *)) {
```
Comment (ned-deily) [inline Modules/_ctypes/malloc_closure.c]: `__builtin_available` again

File: Modules/_ctypes/callbacks.c [hunk 6]
```diff
@@ -408,13 +409,29 @@ CThunkObject *_ctypes_alloc_callback(PyObject *callable,
...
+    {
+#if USING_APPLE_OS_LIBFFI && defined(__arm64__)
+        PyErr_Format(PyExc_NotImplementedError, "ffi_prep_closure_loc() is missing");
+        goto error;
 #else
-    result = ffi_prep_closure_loc(p->pcl_write, &p->cif, closure_fcn,
-                                  p,
-                                  p->pcl_exec);
+        #pragma clang diagnostic push
+        #pragma clang diagnostic ignored "-Wdeprecated-declarations"
```
Comment (ned-deily) [inline Modules/_ctypes/callbacks.c:431]: ditto

File: Modules/_ctypes/callbacks.c [hunk 7]
```diff
@@ -408,13 +409,29 @@ CThunkObject *_ctypes_alloc_callback(PyObject *callable,
...
+        result = ffi_prep_closure(p->pcl_write, &p->cif, closure_fcn, p);
+        #pragma clang diagnostic pop
```
Comment (ned-deily) [inline Modules/_ctypes/callbacks.c:435]: ditto

File: Modules/_ctypes/callbacks.c [hunk 8]
```diff
@@ -408,13 +409,29 @@ CThunkObject *_ctypes_alloc_callback(PyObject *callable,
...
+#endif
+    {
+#if USING_APPLE_OS_LIBFFI && defined(__arm64__)
+        PyErr_Format(PyExc_NotImplementedError, "ffi_prep_closure_loc() is missing");
+        goto error;
 #else
-    result = ffi_prep_closure_loc(p->pcl_write, &p->cif, closure_fcn,
-                                  p,
-                                  p->pcl_exec);
+        #pragma clang diagnostic push
```
Comment (ned-deily) [inline Modules/_ctypes/callbacks.c:430]: We should avoid `-Wunknown-pragmas` warnings when building on other platforms.

**Review by ned-deily — CHANGES_REQUESTED** [review]
As a smoke test, I tried building with `--with-system-ffi` on macOS 10.9 and ran into the issue that `__builtin_available` is apparently not available until `Xcode 9.0 `on `macOS 10.12` and it is a requirement that we can continue to be built on older systems with their supported Xcode versions. I found [a workaround from the curl project](https://github.com/curl/curl/commit/870d849d48a26b8eeb0d4bb1f4655367a4a191ca) that might be usable here if we need to support using the system libffi which would be desirable.

Also we should avoid the potential pragma warnings when building on other platforms.

File: Modules/_ctypes/callbacks.c (hunk 2 above)
Comment (ronaldoussoren) [reply Modules/_ctypes/callbacks.c]: This use of __builtin_available should be safe, it is within a guard of HAVE_FFI_PREP_CLOSURE_LOC and USING_APPLE_OS_LIBFFI: The system version of libffi does not have ffi_prep_closure_loc in the 10.9 SDK (it is introduced in 10.15).

File: Doc/library/ctypes.rst [hunk 9]
```diff
@@ -2545,4 +2545,3 @@ Arrays and pointers
 
         Returns the object to which to pointer points.  Assigning to this
         attribute changes the pointer to point to the assigned object.
-
```
Comment (ronaldoussoren) [inline Doc/library/ctypes.rst]: Unnecessary removal of a blank line (not important)

File: Modules/_ctypes/callbacks.c (hunk 2 above)
Comment (ronaldoussoren) [inline Modules/_ctypes/callbacks.c]: Why test for macOS 11 and not 10.15? According to the header files this API is introduced in 10.15.

File: Modules/_ctypes/callproc.c (hunk 3 above)
Comment (ronaldoussoren) [inline Modules/_ctypes/callproc.c]: Likewise: the headers claim this _var variant is present in macOS 10.15.

File: Modules/_ctypes/callproc.c [hunk 10]
```diff
@@ -835,14 +838,63 @@ static int _call_function_pointer(int flags,
...
-    if (FFI_OK != ffi_prep_cif(&cif,
-                               cc,
-                               argcount,
-                               restype,
-                               atypes)) {
-        PyErr_SetString(PyExc_RuntimeError,
-                        "ffi_prep_cif failed");
-        return -1;
+
+#   if USING_APPLE_OS_LIBFFI
```
Comment (ronaldoussoren) [inline Modules/_ctypes/callproc.c:842]: Proposed change:

#if USING_APPLE_OS_LIBFFI && HAVE_FFI_PREP_CIF_VAR

This way the __builtin_available is not used when building on older macOS versions.

File: Modules/_ctypes/malloc_closure.c (hunk 4 above)
Comment (ronaldoussoren) [reply Modules/_ctypes/malloc_closure.c]: I'm afraid this needs a configure test, ffi_closure_free is not available at all in older SDKs.

File: Modules/_ctypes/malloc_closure.c (hunk 5 above)
Comment (ronaldoussoren) [reply Modules/_ctypes/malloc_closure.c]: same here

File: setup.py [hunk 11]
```diff
@@ -2136,7 +2150,13 @@ def configure_ctypes(self, ext):
...
     def detect_ctypes(self):
         # Thomas Heller's _ctypes module
-        self.use_system_libffi = False
+
+        if (not sysconfig.get_config_var("LIBFFI_INCLUDEDIR") and MACOS and
+            (is_macosx_at_least((10,15)) or '-arch arm64' in sysconfig.get_config_var("CFLAGS"))):
+            self.use_system_libffi = True
+        else:
+            self.use_system_libffi = '--with-system-ffi' in sysconfig.get_config_var("CONFIG_ARGS")
+
```
Comment (ronaldoussoren) [inline setup.py:2102]: I'd prefer to just drop "use_system_libffi" and unconditionally use the system headers.

**Review by ronaldoussoren — COMMENTED** [review]
I've added some comments, and am finally working on this PR on the DTK.

File: Modules/_ctypes/callbacks.c (hunk 2 above)
Comment (lawrence-danna-apple) [reply Modules/_ctypes/callbacks.c]: you're right it should be 10.15

File: setup.py (hunk 11 above)
Comment (lawrence-danna-apple) [reply setup.py:2102]: ok

File: Modules/_ctypes/callbacks.c (hunk 8 above)
Comment (lawrence-danna-apple) [reply Modules/_ctypes/callbacks.c:430]: fixed

**Comment by lawrence-danna-apple** [comment]
I have made the requested changes; please review again

**Comment by lawrence-danna-apple** [comment]
@ned-deily 

Your comments have alerted me to a new problem.    If I understand you correctly, the way you're doing the official builds for python is to build on an old version of Mac OS, with an old version of Xcode, in order to assure compatibility with a broad range of Mac OS versions.

This will work for x86_64, but if you want a universal build that works on both 10.10  and 11.0-arm64, it will not work, becaue the old Xcode doesn't support arm64 for Mac OS

On the other hand, if you build with the new Xcode using MACOSX_DEPLOYMENT_TARGET=10.10, that also does not work on 10.10, because autoconf will discover several functions such as `getentropy` which are available in the SDK, but are not available at runtime on 10.10.   This will cause `-Wunguarded-availability` warnings in the build and an abort at runtime if the symbol is used.

**Comment by maxbelanger** [comment]
@lawrence-danna-apple quite right. This has been an issue for us at Dropbox in recent years. Python's official approach for backwards compatibility is not based on setting the deployment target, which is problematic for use in client-side software. We've fixed this internally by patching `pyconfig.h`, but that's not a great fix (ideally we could do something like weak-link them). 

See https://bugs.python.org/issue31359.

**Comment by lawrence-danna-apple** [comment]
@maxbelanger  @ned-deily 

I opened up a new PR to deal with the deployment target issue.

https://github.com/python/cpython/pull/21576

**Comment by ned-deily** [comment]
Thanks for the PR. We are aware of the issues with supporting multiple levels and architectures  Being able to weaklink from a build on a newer version to an older version has been desirable but it hasn't been a high priority issue. Actively supporting multiple architectures again (something we haven't had to do since the retirement of PPC :) makes this a higher priority issue and @ronaldoussoren is now working on it. I've added him to the reviewers of the new PR (#21577).

File: setup.py [hunk 12]
```diff
@@ -2183,33 +2145,44 @@ def detect_ctypes(self):
...
-            # in /usr/include/ffi
-            ffi_inc_dirs.append('/usr/include/ffi')
-
-        ffi_inc = [sysconfig.get_config_var("LIBFFI_INCLUDEDIR")]
-        if not ffi_inc or ffi_inc[0] == '':
-            ffi_inc = find_file('ffi.h', [], ffi_inc_dirs)
-        if ffi_inc is not None:
-            ffi_h = ffi_inc[0] + '/ffi.h'
+            ext.extra_compile_args.append("-DUSING_APPLE_OS_LIBFFI=1")
+            ffi_in_sdk = os.path.join(macosx_sdk_root(), "usr/include/ffi")
```
Comment (jmroot) [inline setup.py:2154]: Please bear in mind that Python may be built against an external libffi on macOS that is not the OS-provided version. In MacPorts we build against our port of libffi 3.3 on all OS versions.

Comment (lawrence-danna-apple) [reply setup.py:2154]: See previous comment by @ronaldoussoren who prefers to unconditionally use the system's libffi.   I updated this PR accordingly.   Is there any reason not to use the system libffi anymore? I tested going back to 10.10 i think for the other PR adding availability checks.   What's the need to support alternative libffi these days?

Comment (jmroot) [reply setup.py:2154]: Same reasons as any other library: the OS often ships an older version which naturally lacks the features and bug fixes of later versions.

**Comment by ned-deily** [comment]
Thanks for the PR.  It has been included in and superseded by GH-22855.

**Pull request closed without merging**